File: src/languages/sindhi/sindhi.ts
/**
 * Native Sindhi (sd) text phonemizer — canonical IPA. Perso-Arabic (Sindhi) ABJAD → the
 * consonant + LONG-vowel skeleton (incl. the signature IMPLOSIVES ٻɓ ڏɗ ڄʄ ڳɠ and the retroflex series) via a
 * rule g2p; SHORT vowels are usually UNWRITTEN, so a default [ə] stands in (the abjad wall, as for Urdu — folded in
 * the referee eval). ھ aspirates the sonorants + ج/گ (جھ→d͡ʒʰ); و/ي are long vowels after a consonant, glides
 * elsewhere.
 */
import type { Phonemizer } from "../../registry.ts";
import { renderNumber, spellDigits, type NumbersDef } from "../../core/numbers.ts";
import { normalizeSindhi } from "./normalize.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { LATIN_RUN } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { loadTsvMap } from "../../core/loadTsv.ts";
import { applyWeightStress } from "../../core/weightStress.ts";

interface SindhiDef {
    consonants: Record<string, string>;
    aspirateWithHe: Record<string, string>;
    longVowels: Record<string, string>;
    glides: Record<string, string>;
    harakat: Record<string, string>;
    clausePunctuation: Record<string, string>;
    /** Indic lakh/crore number words — see the note in sindhi.jsonc on how each form was sourced. */
    numbers: NumbersDef;
}
const DEF = loadManifest<SindhiDef>(import.meta.url, "sindhi.jsonc");
const CLAUSE_MARK = DEF.clausePunctuation;

export type ForeignPhonemizer = (latin: string) => string;

const HE = "ھ";
// The scan marks its DEFAULT-inserted [ə] with a private-use sentinel so the nasal-assimilation pass can consume
// that ə without touching an ə the writer actually spelled with a harakat. Without this, نَب (explicit fatha)
// lost its written vowel: `nə?(?=[bpɓ])` ate it and the word came out mb. Stripped back to ə at the end of
// phonemizeCore, so the sentinel never escapes this module.
const DEFAULT_SCHWA = "\uE000";
const NOON_GHUNNA = "ں"; // nasalization
const isConsonant = (c: string): boolean => c in DEF.consonants;
const isVowelLetter = (c: string): boolean => c in DEF.longVowels;

/** Scan a Sindhi (Perso-Arabic) word → IPA. Consonants + long vowels are recoverable; short vowels default to [ə].
 *
 *  `vocalized`: the word is FULLY harakat-marked (a diacritized dictionary headword rather than running text).
 *  Then the diacritics are authoritative and the default-[ə] insertion must be OFF — an unmarked consonant cluster
 *  means there is NO vowel there, not an unwritten one. Reading اُستادُ with insertion on gives ʊsət̪aːd̪ʊ; the
 *  attested form is ʊst̪aːd̪ʊ. */
function scan(word: string, vocalized = false): string {
    const s = [...word];
    const out: string[] = [];
    let prevNucleus = false; // was the last emitted unit a vowel nucleus?
    for (let i = 0; i < s.length; ) {
        const c = s[i]!;
        // harakat → explicit short vowel
        if (DEF.harakat[c]) {
            out.push(DEF.harakat[c]!);
            prevNucleus = true;
            i++;
            continue;
        }
        // ں → nasalize the preceding vowel
        if (c === NOON_GHUNNA) {
            if (prevNucleus) out[out.length - 1] += "̃";
            i++;
            continue;
        }
        // long-vowel / glide letters: a long vowel after a consonant, a glide word-initially or in hiatus
        if (isVowelLetter(c)) {
            if (out.length === 0) {
                // word-initial vowel carrier: آ (madda) is [aː]; bare ا is a short [ə] carrier; و/ي → glide.
                // If a harakat FOLLOWS the bare alif it is the carrier's vowel (اِجازت ɪd͡ʒaːzət̪ə, اُستاد ʊst̪aːd̪ʊ) —
                // emit nothing here and let the harakat supply it, or the two stack into a spurious əi/əu.
                if (c === "ا" && DEF.harakat[s[i + 1] ?? ""]) {
                    prevNucleus = false;
                    i++;
                    continue;
                }
                out.push(c === "آ" ? "aː" : c === "ا" ? "ə" : (DEF.glides[c] ?? DEF.longVowels[c]!));
                prevNucleus = c === "و" || c === "ي" || c === "ی" || c === "ے" ? false : true;
            } else if (prevNucleus) {
                out.push(DEF.glides[c] ?? DEF.longVowels[c]!);
                prevNucleus = false;
            } else {
                out.push(DEF.longVowels[c]!);
                prevNucleus = true;
            }
            i++;
            continue;
        }
        // ئ/ؤ are hamza SEATS (not a glottal stop): they carry a hiatus vowel — emit nothing but break the glide so
        // a following ي/و is read as a full vowel (آئينو→aːiːnoː, not aːjnoː).
        if (c === "ئ" || c === "ؤ") {
            prevNucleus = false;
            i++;
            continue;
        }
        // ع is usually SILENT / a vowel modifier in Sindhi (تعليم→t̪əliːm), not a full [ʔ] — skip it.
        if (c === "ع") {
            i++;
            continue;
        }
        // word-final ه/ہ/ح is a silent vowel-carrier (ٻه→ɓə, روح→ruh→ru), not [h]
        if ((c === "ه" || c === "ہ" || c === "ح") && i === s.length - 1) {
            i++;
            continue;
        }
        // consonants; do-chashmi ھ (U+06BE) after ج/گ/a sonorant → aspiration (جھ→d͡ʒʰ, لھ→lʰ). Plain ه (U+0647)
        // is the /h/ CONSONANT, NOT an aspiration marker — the letters are contrastive (نه→nə, مهينو→məhiːnoː,
        // NOT nʰə/mʰiːnoː). Words that spell aspiration with plain ه (گهوڙو) are orthographic variants that collide
        // with real /h/ and can't be disambiguated from the letter alone (the abjad ambiguity — a lexicon tail).
        if (isConsonant(c)) {
            if (s[i + 1] === HE && DEF.aspirateWithHe[c]) {
                out.push(DEF.aspirateWithHe[c]!);
                i += 2;
            } else {
                out.push(DEF.consonants[c]!);
                i++;
            }
            prevNucleus = false;
            // insert a default short [ə] when no vowel is written before the next consonant / word-end (abjad).
            const nx = s[i] ?? "";
            if (!vocalized && (nx === "" || (isConsonant(nx) && !(nx === HE)))) {
                out.push(DEFAULT_SCHWA);
                prevNucleus = true;
            }
            continue;
        }
        i++; // unknown → skip
    }
    return out.join("");
}

/** Rule g2p: consonant + long-vowel skeleton with default-[ə] short vowels, then homorganic nasal assimilation
 *  before a velar → ŋ (انگ→əŋɡ), labial → m (انب→əmb), retroflex → ɳ (آنڊو→aːɳɖo), palatal → ɲ (پنج→pəɲd͡ʒ),
 *  dental → plain n (سنڌ→sənd̪ʰ).
 *
 *  Each rule also CONSUMES the default [ə] the abjad scan inserted between the nasal and the stop. A homorganic
 *  nasal + stop is a single tautosyllabic cluster in Sindhi — it is never broken by a vowel — so the scan's
 *  blanket "consonant before consonant → ə" is wrong here. The lexicon goldens show this directly: انب is əmb
 *  (not əməb) and سنڌي is sɪndʱiː (not sɪnədʱiː). Measured over the 539-word lexicon, dropping the ə here fixes
 *  12 of the 53 places where the rule path split a cluster the attested form keeps (53 → 41).
 *
 *  The other 41 (sC- st/sp/sk, obstruent+liquid kɾ/kl/pɾ, and assorted codas lm/ɾs/kt) are deliberately left
 *  alone: unlike the homorganic nasal they are not categorically un-splittable in Sindhi, and the lexicon does
 *  not carry enough same-shape pairs to tell a real cluster from a genuine epenthesis word-by-word. Re-measured
 *  later on the full 9,920-word lexicon, only 8.9% of ALL default-ə insertions are wrong, so this is a narrow
 *  correction rather than a large one. */
function phonemizeCore(word: string, vocalized = false): string {
    return scan(word, vocalized)
        .replace(/n\uE000?(?=(?:kʰ|[kɡxɠ]))/gu, "ŋ")
        .replace(/n\uE000?(?=[bpɓ])/gu, "m")
        .replace(/n\uE000?(?=[ʈɖɳɽ])/gu, "ɳ")
        .replace(/n\uE000?(?=(?:d͡ʒ|t͡ʃ|ʄ))/gu, "ɲ")
        .replace(/n\uE000(?=t̪|d̪)/gu, "n")
        .replace(/\uE000/gu, "ə")
        .normalize("NFC");
}

// SHORT-VOWEL restoration lexicon (sindhi-lexicon.tsv): bare word → voweled IPA, mined from kaikki Sindhi
// (Wiktionary, CC BY-SA). The Perso-Arabic abjad leaves short vowels unwritten, so the rule g2p defaults every
// one to [ə] (زبان → zəbaːnə); this restores the attested Sindhi short vowels (zʊbaːnə), the Urdu restoreHarakat
// pattern. The referee eval FOLDS short vowels (abjad wall) so it scores the RULE path (phonemizeWordRules,
// non-circular); this lexicon is a SHIPPED refinement — its value is the correct vocalization for TTS.
let LEX: ReadonlyMap<string, string> | undefined;
const lexicon = (): ReadonlyMap<string, string> =>
    (LEX ??= loadTsvMap(import.meta.url, "sindhi-lexicon.tsv", undefined, { optional: true }));

/** One Sindhi word → canonical IPA, SHIPPED path (rule g2p + the kaikki short-vowel restoration lexicon).
 *  Lexicon entries are stored UNSTRESSED (as for Urdu), so weight stress is applied at lookup — one stress
 *  policy for both the lexicon and the rule path. */
export function phonemizeWord(word: string): string {
    return stress(lexicon().get(word) ?? phonemizeCore(word));
}

/** One Sindhi word → canonical IPA, RULE-ONLY (no lexicon) — the non-circular signal for the referee eval. */
export function phonemizeWordRules(word: string): string {
    return stress(phonemizeCore(word));
}

/** Quantity-sensitive word stress — the shared Indo-Aryan weight rule already used by hi/ur/pa
 *  (rightmost superheavy, else rightmost non-final heavy, else initial). Sindhi is the same stress family;
 *  this module simply never had the layer, so every word came out unstressed. */
function stress(ipa: string): string {
    return applyWeightStress(ipa).normalize("NFC");
}

const SD_WORD = "ء-ٟٮ-ۿ";
// Latin runs and digits BOTH route through `foreign` (the English phonemizer, wired in the registry — the
// ur/hi pattern). They used to be dropped outright: Latin matched no group at all, and digits emitted ""
// because createSindhi() never passed a foreign phonemizer — 7% of FLEURS sd_in tokens silently vanished.
/** A digit run → Sindhi number words → IPA through this engine's own g2p. Indic lakh/crore grouping. */
function number(digits: string): string {
    const n = Number(digits);
    // ⚠ ABOVE 2^53 THIS USED TO RETURN "" AND THE NUMBER VANISHED — the same silent deletion this file's
    // header records for the pre-composer era, reintroduced at the top of the range. Refusing to COMPOSE is
    // right (the float has lost the low digits); the guard simply had no else. Digit-at-a-time out of the
    // manifest's own unit words invents nothing. Above 2^53 the reading is a digit string, not a quantity.
    if (!Number.isSafeInteger(n)) return spellDigits(digits, DEF.numbers, (w) => phonemizeWordWith(w));
    return renderNumber(n, DEF.numbers, (w) => phonemizeWordWith(w));
}

// The foreign arm is `LATIN_RUN`, ALL of Latin plus marks — not `[A-Za-z]+`, which ended the token at a
// diacritic and left that letter to be read as an English letter name (`Cañitas` → *ka ˈɛn ˈitas*). This
// engine ROUTES a foreign word to the injected reader, so widening the class is the whole fix.
const TOKEN = new RegExp(`([${SD_WORD}]+)|(${LATIN_RUN})|(\\d+)|([۔؟،؛.?,])`, "gu");

/** Resolve an OOV word to IPA. Consulted BETWEEN the lexicon and the rule engine (lexicon → oovOverride →
 *  rules); used only by the async neural path (`sindhiNeural.ts`), so the sync engine is unchanged. */
export type OovResolver = (word: string) => string | undefined;

class SindhiPhonemizer implements Phonemizer {
    constructor(private foreign?: ForeignPhonemizer) {}
    text(rawInput: string, oovOverride?: OovResolver): string {
        // everything the g2p cannot read is rewritten to Sindhi words FIRST — see normalize.ts.
        // This layer was pointless until the number composer landed, since every rule there
        // produces digits for that composer to read.
        return assembleClauses(normalizeSindhi(rawInput), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWordWith(m[1], oovOverride));
            // A LATIN run goes to the foreign phonemizer; a DIGIT run does NOT. Sending digits there
            // meant every numeral in Sindhi text was spoken in English.
            else if (m[2]) sink.emit(this.foreign ? this.foreign(m[2]) : "");
            else if (m[3]) sink.emit(number(m[3]));
            else if (m[4]) {
                const mk = CLAUSE_MARK[m[4]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Lexicon → oovOverride → rules. The lexicon still wins any word it covers; the override only fills the tail. */
function phonemizeWordWith(word: string, oov?: OovResolver): string {
    const hit = lexicon().get(word);
    if (hit) return stress(hit);
    const o = oov?.(word);
    return o ? stress(o) : stress(phonemizeCore(word));
}

/** True when the shipped lexicon covers `word` — the neural path uses this to skip covered words.
 *  Deliberately the SAME lookup the engine performs (`lexicon().get(word)`, no normalisation): if the two
 *  disagreed, a word could be judged "covered" here, skipped by the tagger, and then miss in the engine —
 *  silently falling through to default-ə instead of the neural reading. */
export function sindhiLexiconHas(word: string): boolean {
    return lexicon().has(word);
}

/** Build the Sindhi phonemizer. `foreign` reads embedded Latin words and digit runs (wired to English in
 *  the registry, as for ur/hi — code-switching is normal in Sindhi text; silence is not). */
export function createSindhi(foreign?: ForeignPhonemizer): Phonemizer {
    return new SindhiPhonemizer(foreign);
}

/** Build the Sindhi engine with a per-call `oovOverride` hook (the async neural path). */
export function createSindhiEngine(): { text: (input: string, oov?: OovResolver) => string } {
    return new SindhiPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Sindhi/Sindhi.cs
/**
 * Native Sindhi (sd) text phonemizer — canonical IPA. Perso-Arabic (Sindhi) ABJAD → the consonant +
 * LONG-vowel skeleton (the implosives ٻɓ ڏɗ ڄʄ ڳɠ and the retroflex series included) via a rule g2p; the
 * usually-unwritten SHORT vowels default to [ə], restored on covered words by the shipped lexicon.
 * Ported from src/languages/sindhi/sindhi.ts — see that file for the corpus evidence.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Sindhi;

/** Resolve an OOV word to IPA — consulted BETWEEN the lexicon and the rule engine; null for "no opinion".
 *  Used only by the async neural path (SindhiNeural.cs), so the sync engine is unchanged. */
public delegate string? OovResolver(string word);

public sealed class SindhiPhonemizer : ILanguage
{
    private static SindhiDef DEF => Manifest.DEF;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;

    private const string HE = "ھ";

    /** The scan marks its DEFAULT-inserted [ə] with a private-use sentinel so the nasal-assimilation pass
     *  consumes only that ə and never one the writer spelled with a harakat. Stripped at the end of
     *  PhonemizeCore, so it never escapes this module. */
    private const string DEFAULT_SCHWA = "\uE000";
    private const string NOON_GHUNNA = "ں"; // nasalization

    private static bool IsConsonant(string c) => DEF.Consonants.ContainsKey(c);
    private static bool IsVowelLetter(string c) => DEF.LongVowels.ContainsKey(c);

    /** JS `if (DEF.harakat[c])` — a missing key AND an empty string are both falsy. */
    private static string? Harakat(string c) =>
        DEF.Harakat.TryGetValue(c, out var v) && v.Length > 0 ? v : null;

    /** Scan a Sindhi (Perso-Arabic) word → IPA. Consonants + long vowels are recoverable; short vowels
     *  default to [ə]. `vocalized`: the word is FULLY harakat-marked, so the diacritics are authoritative
     *  and the default-[ə] insertion is OFF. */
    private static string Scan(string word, bool vocalized = false)
    {
        var s = Js.CodePoints(word);
        var outp = new List<string>();
        var prevNucleus = false; // was the last emitted unit a vowel nucleus?
        for (var i = 0; i < s.Count;)
        {
            var c = s[i];
            var h = Harakat(c);
            if (h is not null)
            {
                outp.Add(h);
                prevNucleus = true;
                i++;
                continue;
            }
            if (c == NOON_GHUNNA)
            {
                if (prevNucleus) outp[^1] += "̃";
                i++;
                continue;
            }
            // long-vowel / glide letters: a long vowel after a consonant, a glide word-initially or in hiatus
            if (IsVowelLetter(c))
            {
                if (outp.Count == 0)
                {
                    // A harakat FOLLOWING a bare alif is the carrier's own vowel — emit nothing here, or the
                    // two stack into a spurious əi/əu.
                    if (c == "ا" && Harakat(i + 1 < s.Count ? s[i + 1] : "") is not null)
                    {
                        prevNucleus = false;
                        i++;
                        continue;
                    }
                    outp.Add(c == "آ" ? "aː"
                        : c == "ا" ? "ə"
                        : DEF.Glides.TryGetValue(c, out var gl) ? gl : DEF.LongVowels[c]);
                    prevNucleus = !(c == "و" || c == "ي" || c == "ی" || c == "ے");
                }
                else if (prevNucleus)
                {
                    outp.Add(DEF.Glides.TryGetValue(c, out var gl2) ? gl2 : DEF.LongVowels[c]);
                    prevNucleus = false;
                }
                else
                {
                    outp.Add(DEF.LongVowels[c]);
                    prevNucleus = true;
                }
                i++;
                continue;
            }
            // ئ/ؤ are hamza SEATS: emit nothing, but break the glide so a following ي/و is a full vowel.
            if (c == "ئ" || c == "ؤ")
            {
                prevNucleus = false;
                i++;
                continue;
            }
            if (c == "ع") // usually SILENT / a vowel modifier in Sindhi, not a full [ʔ]
            {
                i++;
                continue;
            }
            if ((c == "ه" || c == "ہ" || c == "ح") && i == s.Count - 1) // word-final silent vowel-carrier
            {
                i++;
                continue;
            }
            if (IsConsonant(c))
            {
                if (i + 1 < s.Count && s[i + 1] == HE && DEF.AspirateWithHe.TryGetValue(c, out var asp))
                {
                    outp.Add(asp);
                    i += 2;
                }
                else
                {
                    outp.Add(DEF.Consonants[c]);
                    i++;
                }
                prevNucleus = false;
                // insert a default short [ə] when no vowel is written before the next consonant / word-end
                var nx = i < s.Count ? s[i] : "";
                if (!vocalized && (nx == "" || (IsConsonant(nx) && nx != HE)))
                {
                    outp.Add(DEFAULT_SCHWA);
                    prevNucleus = true;
                }
                continue;
            }
            i++; // unknown → skip
        }
        return string.Concat(outp);
    }

    private static readonly JsRe N_VELAR = JsRegex.Compile("n\uE000?(?=(?:kʰ|[kɡxɠ]))", "gu");
    private static readonly JsRe N_LABIAL = JsRegex.Compile("n\uE000?(?=[bpɓ])", "gu");
    private static readonly JsRe N_RETROFLEX = JsRegex.Compile("n\uE000?(?=[ʈɖɳɽ])", "gu");
    private static readonly JsRe N_PALATAL = JsRegex.Compile("n\uE000?(?=(?:d͡ʒ|t͡ʃ|ʄ))", "gu");
    private static readonly JsRe N_DENTAL = JsRegex.Compile("n\uE000(?=t̪|d̪)", "gu");
    private static readonly JsRe SENTINEL = JsRegex.Compile("\uE000", "gu");

    /** Rule g2p: the consonant + long-vowel skeleton with default-[ə] short vowels, then homorganic nasal
     *  assimilation — each rule also CONSUMING the default [ə] the abjad scan inserted inside the cluster. */
    private static string PhonemizeCore(string word, bool vocalized = false)
    {
        var s = Scan(word, vocalized);
        s = N_VELAR.Replace(s, "ŋ");
        s = N_LABIAL.Replace(s, "m");
        s = N_RETROFLEX.Replace(s, "ɳ");
        s = N_PALATAL.Replace(s, "ɲ");
        s = N_DENTAL.Replace(s, "n");
        s = SENTINEL.Replace(s, "ə");
        return s.Normalize(NormalizationForm.FormC);
    }

    /** SHORT-VOWEL restoration lexicon (sindhi-lexicon.tsv): bare word → voweled IPA, mined from kaikki. */
    private static IReadOnlyDictionary<string, string>? LEX;
    private static IReadOnlyDictionary<string, string> Lexicon() =>
        LEX ??= LoadTsv.LoadTsvMap("languages/sindhi", "sindhi-lexicon.tsv", optional: true);

    /** One Sindhi word → canonical IPA, SHIPPED path (rule g2p + the kaikki short-vowel restoration lexicon).
     *  Lexicon entries are stored UNSTRESSED, so weight stress is applied at lookup. */
    public static string PhonemizeWord(string word) =>
        Stress(Lexicon().TryGetValue(word, out var hit) ? hit : PhonemizeCore(word));

    /** One Sindhi word → canonical IPA, RULE-ONLY (no lexicon) — the non-circular signal for the referee eval. */
    public static string PhonemizeWordRules(string word) => Stress(PhonemizeCore(word));

    /** Quantity-sensitive word stress — the shared Indo-Aryan weight rule already used by hi/ur/pa. */
    private static string Stress(string ipa) =>
        WeightStress.ApplyWeightStress(ipa).Normalize(NormalizationForm.FormC);

    private const string SD_WORD = "ء-ٟٮ-ۿ";

    /** A digit run → Sindhi number words → IPA through this engine's own g2p. Indic lakh/crore grouping. */
    private static string Number(string digits)
    {
        var n = Js.Number(digits);
        // ⚠ Above 2^53 the float has lost the low digits, so composing a quantity is refused; the reading
        // falls back to a digit string rather than vanishing.
        if (!(double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d))
            return Core.Numbers.SpellDigits(digits, DEF.Numbers, w => PhonemizeWordWith(w));
        return Core.Numbers.RenderNumber(n, DEF.Numbers, w => PhonemizeWordWith(w));
    }

    private static readonly JsRe TOKEN =
        JsRegex.Compile($"([{SD_WORD}]+)|({HostWord.LATIN_RUN})|(\\d+)|([۔؟،؛.?,])", "gu");

    private readonly Func<string, string>? _foreign;
    private SindhiPhonemizer(Func<string, string>? foreign = null) => _foreign = foreign;

    public string Text(string rawInput) => Text(rawInput, null);

    public string Text(string rawInput, OovResolver? oovOverride)
    {
        // everything the g2p cannot read is rewritten to Sindhi words FIRST — see Normalize.cs.
        return Clauses.AssembleClauses(Normalize.NormalizeSindhi(rawInput), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                sink.Emit(PhonemizeWordWith(m.Groups[1].Value, oovOverride));
            // A LATIN run goes to the foreign phonemizer; a DIGIT run does NOT.
            else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                sink.Emit(_foreign is not null ? _foreign(m.Groups[2].Value) : "");
            else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                sink.Emit(Number(m.Groups[3].Value));
            else if (m.Groups[4].Success && m.Groups[4].Value.Length > 0)
            {
                var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[4].Value);
                if (!string.IsNullOrEmpty(mk)) sink.Pause(mk);
            }
        });
    }

    /** Lexicon → oovOverride → rules. The lexicon still wins any word it covers. */
    private static string PhonemizeWordWith(string word, OovResolver? oov = null)
    {
        if (Lexicon().TryGetValue(word, out var hit) && hit.Length > 0) return Stress(hit);
        var o = oov?.Invoke(word);
        return !string.IsNullOrEmpty(o) ? Stress(o) : Stress(PhonemizeCore(word));
    }

    /** True when the shipped lexicon covers `word` — the neural path uses this to skip covered words.
     *  Deliberately the SAME lookup the engine performs (no normalisation). */
    public static bool SindhiLexiconHas(string word) => Lexicon().ContainsKey(word);

    /** Build the Sindhi phonemizer. `foreign` reads embedded Latin words and digit runs. */
    public static SindhiPhonemizer CreateSindhi(Func<string, string>? foreign = null) => new(foreign);

    /** Build the Sindhi engine with a per-call `oovOverride` hook (the async neural path). */
    public static SindhiPhonemizer CreateSindhiEngine() => new();

    internal static void RegisterSelf() => Registry.Register("sindhi", () => CreateSindhi(Registry.ReadAsEnglish));
}
